File: src/commands/start.ts
import { appendFile, mkdtemp, readFile, rm, writeFile } from 'node:fs/promises'
import { tmpdir } from 'node:os'
import { join } from 'node:path'

import { moveCopyCursor } from '../core/action'
import { capturePane, fitCaptureToHeight } from '../core/capture'
import { createStyledDisplayCells } from '../core/width'
import {
  createDaemonSocketPath,
  ensureDaemon,
  resolveCliEntrypoint,
} from './runtime'
import {
  createTmuxClient,
  displayPopup,
  enterCopyMode,
  focusClientPane,
  getPaneStartContext,
  getPaneBorderLines,
  listWindowPanes,
  type PaneBorderLines,
  type DisplayPopupOptions,
  type PaneStartContext,
} from '../infra/tmux'
import type { InputResult, InputState, PaneSnapshot } from '../types'

type StartScope = 'current' | 'all'

type ParsedStartArgs = {
  scope: StartScope
  paneId: string
  clientTty: string
}

type PopupState = {
  currentPath: string
  state: InputState
  x?: string | number
  y?: string | number
}

type DebugLogPayload = {
  event: string
  [key: string]: unknown
}

const writeDebugLog = async (payload: DebugLogPayload): Promise<void> => {
  const debugLogPath = process.env.TMUX_FUZZY_MOTION_DEBUG_LOG
  if (!debugLogPath) {
    return
  }

  const line = JSON.stringify({
    timestamp: new Date().toISOString(),
    ...payload,
  })

  try {
    await appendFile(debugLogPath, `${line}\n`, 'utf8')
  } catch {
    // Ignore debug logging failures.
  }
}

const buildPopupRelativePosition = (
  axis: 'x' | 'y',
  targetOrigin: number,
): string => {
  if (axis === 'x') {
    return `#{e|+|:#{popup_pane_left},#{e|-|:${targetOrigin},#{pane_left}}}`
  }

  const topOffset =
    '#{?#{==:#{status-position},top},#{e|-|:#{client_height},#{window_height}},0}'
  const desiredTop = `#{e|+|:#{e|-|:${targetOrigin},#{window_offset_y}},${topOffset}}`

  return `#{e|+|:#{popup_height},${desiredTop}}`
}

const buildPopupCommand = (
  stateFile: string,
  resultFile: string,
  socketPath: string,
): string[] => [
  process.execPath,
  resolveCliEntrypoint(),
  'popup',
  '--state-file',
  stateFile,
  '--result-file',
  resultFile,
  '--socket',
  socketPath,
]

const readResult = async (resultFile: string): Promise<InputResult> => {
  try {
    return JSON.parse(await readFile(resultFile, 'utf8')) as InputResult
  } catch (error) {
    throw new Error('tmux-fuzzy-motion: popup did not produce result', {
      cause: error,
    })
  }
}

const parseStartArgs = (args: string[]): ParsedStartArgs => {
  let scope: StartScope = 'current'
  const positional: string[] = []

  for (let index = 0; index < args.length; index += 1) {
    const value = args[index]
    if (value === '--scope') {
      const nextScope = args[index + 1]
      if (nextScope === 'current' || nextScope === 'all') {
        scope = nextScope
      }
      index += 1
      continue
    }

    positional.push(value ?? '')
  }

  return {
    scope,
    paneId: positional[0] ?? '',
    clientTty: positional[1] ?? '',
  }
}

const createBlankRow = (width: number): string[] =>
  Array.from({ length: width }, () => ' ')

const BORDER_SETS: Record<
  Exclude<PaneBorderLines, 'number'>,
  { horizontal: string; intersection: string; vertical: string }
> = {
  single: {
    vertical: '│',
    horizontal: '─',
    intersection: '┼',
  },
  double: {
    vertical: '║',
    horizontal: '═',
    intersection: '╬',
  },
  heavy: {
    vertical: '┃',
    horizontal: '━',
    intersection: '╋',
  },
  simple: {
    vertical: '|',
    horizontal: '-',
    intersection: '+',
  },
  spaces: {
    vertical: ' ',
    horizontal: ' ',
    intersection: ' ',
  },
}

const createOccupancyGrid = (
  panes: Pick<PaneSnapshot, 'left' | 'top' | 'width' | 'height'>[],
  width: number,
  height: number,
): boolean[][] => {
  const occupied = Array.from({ length: height }, () =>
    Array.from({ length: width }, () => false),
  )

  for (const pane of panes) {
    for (let row = pane.top; row < pane.top + pane.height; row += 1) {
      const line = occupied[row]
      if (!line) {
        continue
      }

      for (
        let column = pane.left;
        column < pane.left + pane.width;
        column += 1
      ) {
        if (column >= 0 && column < line.length) {
          line[column] = true
        }
      }
    }
  }

  return occupied
}

const resolveBorderSet = (
  borderLines: PaneBorderLines,
): { horizontal: string; intersection: string; vertical: string } =>
  borderLines === 'number' ? BORDER_SETS.simple : BORDER_SETS[borderLines]

const drawPaneBorders = (
  rows: string[][],
  occupied: boolean[][],
  borderLines: PaneBorderLines,
): void => {
  const borderSet = resolveBorderSet(borderLines)

  for (let rowIndex = 0; rowIndex < rows.length; rowIndex += 1) {
    const row = rows[rowIndex]
    const occupiedRow = occupied[rowIndex]
    if (!row || !occupiedRow) {
      continue
    }

    for (let columnIndex = 0; columnIndex < row.length; columnIndex += 1) {
      if (occupiedRow[columnIndex]) {
        continue
      }

      const left = occupiedRow[columnIndex - 1] ?? false
      const right = occupiedRow[columnIndex + 1] ?? false
      const top = occupied[rowIndex - 1]?.[columnIndex] ?? false
      const bottom = occupied[rowIndex + 1]?.[columnIndex] ?? false
      const hasVertical = left || right
      const hasHorizontal = top || bottom

      if (hasVertical && hasHorizontal) {
        row[columnIndex] = borderSet.intersection
      } else if (hasVertical) {
        row[columnIndex] = borderSet.vertical
      } else if (hasHorizontal) {
        row[columnIndex] = borderSet.horizontal
      }
    }
  }
}

const composeDisplayLines = (
  panes: Pick<
    PaneSnapshot,
    'displayLines' | 'left' | 'top' | 'width' | 'height'
  >[],
  width: number,
  height: number,
  borderLines: PaneBorderLines,
): string[] => {
  const rows = Array.from({ length: height }, () => createBlankRow(width))
  const occupied = createOccupancyGrid(panes, width, height)

  for (const pane of panes) {
    pane.displayLines.forEach((line, lineIndex) => {
      const row = rows[pane.top + lineIndex]
      if (!row) {
        return
      }

      const cells = createStyledDisplayCells(line)
      cells.forEach((cell, cellIndex) => {
        const column = pane.left + cellIndex
        if (column < 0 || column >= row.length) {
          return
        }

        row[column] = cell
      })
    })
  }

  drawPaneBorders(rows, occupied, borderLines)

  return rows.map((row) => row.join(''))
}

const buildCurrentState = async (
  tmux: ReturnType<typeof createTmuxClient>,
  pane: PaneStartContext,
  paneId: string,
  clientTty: string,
): Promise<PopupState> => {
  if (!pane.inCopyMode) {
    await enterCopyMode(tmux, paneId)
  }
  const capture = fitCaptureToHeight(
    await capturePane(tmux, paneId),
    pane.height,
  )

  return {
    currentPath: pane.currentPath,
    state: {
      scope: 'current',
      paneId,
      clientTty,
      displayLines: capture.displayLines,
      plainLines: capture.lines,
      width: pane.width,
      height: pane.height,
    },
  }
}

const buildAllPaneState = async (
  tmux: ReturnType<typeof createTmuxClient>,
  pane: PaneStartContext,
  paneId: string,
  clientTty: string,
): Promise<PopupState> => {
  const panes = await listWindowPanes(tmux, paneId)
  const borderLines = await getPaneBorderLines(tmux, paneId)
  const bounds = panes.reduce(
    (accumulator, item) => ({
      left: Math.min(accumulator.left, item.left),
      top: Math.min(accumulator.top, item.top),
      right: Math.max(accumulator.right, item.left + item.width),
      bottom: Math.max(accumulator.bottom, item.top + item.height),
    }),
    {
      left: Number.POSITIVE_INFINITY,
      top: Number.POSITIVE_INFINITY,
      right: Number.NEGATIVE_INFINITY,
      bottom: Number.NEGATIVE_INFINITY,
    },
  )
  const snapshots: PaneSnapshot[] = []

  for (const item of panes) {
    const capture = fitCaptureToHeight(
      await capturePane(tmux, item.paneId),
      item.height,
    )
    snapshots.push({
      paneId: item.paneId,
      inCopyMode: item.inCopyMode,
      width: item.width,
      height: item.height,
      left: item.left - bounds.left,
      top: item.top - bounds.top,
      plainLines: capture.lines,
      displayLines: capture.displayLines,
    })
  }

  const width = Math.max(0, bounds.right - bounds.left)
  const height = Math.max(0, bounds.bottom - bounds.top)
  const x = buildPopupRelativePosition('x', bounds.left)
  const y = buildPopupRelativePosition('y', bounds.top)

  await writeDebugLog({
    event: 'start.build-all-pane-state',
    paneId,
    clientTty,
    targetPane: {
      paneId: pane.paneId,
      width: pane.width,
      height: pane.height,
      inCopyMode: pane.inCopyMode,
      currentPath: pane.currentPath,
    },
    bounds,
    size: {
      width,
      height,
    },
    popupPosition: {
      x,
      y,
    },
    panes: panes.map((item) => ({
      paneId: item.paneId,
      left: item.left,
      top: item.top,
      width: item.width,
      height: item.height,
      active: item.active,
      inCopyMode: item.inCopyMode,
      currentPath: item.currentPath,
    })),
  })

  return {
    currentPath: pane.currentPath,
    x,
    y,
    state: {
      scope: 'all',
      paneId,
      clientTty,
      displayLines: composeDisplayLines(snapshots, width, height, borderLines),
      panes: snapshots,
      width,
      height,
    },
  }
}

export const runStart = async (args: string[]): Promise<number> => {
  const { scope, paneId, clientTty } = parseStartArgs(args)

  if (!process.env.TMUX) {
    console.error('tmux-fuzzy-motion: must be run inside tmux')
    return 2
  }

  if (!paneId) {
    console.error('tmux-fuzzy-motion: pane not found')
    return 2
  }

  if (!clientTty) {
    console.error('tmux-fuzzy-motion: client not found')
    return 2
  }

  const tmux = createTmuxClient()
  const tempDir = await mkdtemp(join(tmpdir(), 'tmux-fuzzy-motion-'))
  const stateFile = join(tempDir, 'state.json')
  const resultFile = join(tempDir, 'result.json')
  const socketPath = createDaemonSocketPath()

  try {
    const pane = await getPaneStartContext(tmux, paneId)
    await focusClientPane(tmux, paneId, clientTty)
    const popupState =
      scope === 'all'
        ? await buildAllPaneState(tmux, pane, paneId, clientTty)
        : await buildCurrentState(tmux, pane, paneId, clientTty)
    const state = popupState.state

    await writeFile(stateFile, JSON.stringify(state), 'utf8')
    await ensureDaemon(socketPath)
    const popupOptions: DisplayPopupOptions = {
      command: buildPopupCommand(stateFile, resultFile, socketPath),
      currentPath: popupState.currentPath,
      height: state.height,
      targetClient: clientTty,
      targetPane: paneId,
      width: state.width,
    }
    if (popupState.x !== undefined) {
      popupOptions.x = popupState.x
    }
    if (popupState.y !== undefined) {
      popupOptions.y = popupState.y
    }

    await writeDebugLog({
      event: 'start.display-popup',
      scope,
      paneId,
      clientTty,
      popupOptions: {
        targetPane: popupOptions.targetPane,
        targetClient: popupOptions.targetClient,
        currentPath: popupOptions.currentPath,
        width: popupOptions.width,
        height: popupOptions.height,
        x: popupOptions.x ?? '#{popup_pane_left}',
        y: popupOptions.y ?? '#{popup_pane_top}',
      },
    })

    await displayPopup(tmux, popupOptions)

    const result = await readResult(resultFile)
    if (result.status === 'selected') {
      const targetPaneId = result.target.paneId ?? paneId
      await tmux.runQuiet(['select-pane', '-t', targetPaneId])
      if (
        state.scope === 'all' &&
        !state.panes.some(
          (targetPane) =>
            targetPane.paneId === targetPaneId && targetPane.inCopyMode,
        )
      ) {
        await enterCopyMode(tmux, targetPaneId)
      }
      await moveCopyCursor(tmux, targetPaneId, result.target)
    }

    return 0
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error)
    console.error(message)
    return message.startsWith('tmux-fuzzy-motion:') ? 2 : 1
  } finally {
    await rm(tempDir, { recursive: true, force: true })
  }
}

File: src/start.rs
use std::env;
use std::fs;
use std::thread;

use anyhow::{Result, anyhow, bail};
use tempfile::tempdir;

use crate::action::move_copy_cursor;
use crate::capture::{capture_pane, fit_capture_to_height};
use crate::tmux::{
    enter_copy_mode, get_pane_start_context, list_window_panes, s, tmux, tmux_quiet,
};
use crate::types::{InputResult, InputState, PaneSnapshot, PopupState, Scope, ScopeArg};
use crate::width::compact_styled_display_cells;

pub fn run_start(scope: ScopeArg, pane_id: String, client_tty: String) -> Result<()> {
    if env::var_os("TMUX").is_none() {
        bail!("tmux-fuzzy-motion: must be run inside tmux");
    }
    let temp_dir = tempdir()?;
    let state_file = temp_dir.path().join("state.json");
    let result_file = temp_dir.path().join("result.json");
    let popup_state = match scope {
        ScopeArg::Current => build_current_state(&pane_id, &client_tty)?,
        ScopeArg::All => build_all_pane_state(&pane_id, &client_tty)?,
    };
    fs::write(&state_file, serde_json::to_vec(&popup_state.state)?)?;

    let exe = env::current_exe()?;
    let mut command = vec![
        "display-popup".to_string(),
        "-E".to_string(),
        "-B".to_string(),
        "-c".to_string(),
        client_tty.clone(),
        "-t".to_string(),
        pane_id.clone(),
        "-d".to_string(),
        popup_state.current_path.clone(),
        "-x".to_string(),
        popup_state
            .x
            .unwrap_or_else(|| "#{popup_pane_left}".to_string()),
        "-y".to_string(),
        popup_state
            .y
            .unwrap_or_else(|| "#{popup_pane_top}".to_string()),
        "-w".to_string(),
        popup_state.state.width.to_string(),
        "-h".to_string(),
        popup_state.state.height.to_string(),
        exe.to_string_lossy().to_string(),
        "popup".to_string(),
        "--state-file".to_string(),
        state_file.to_string_lossy().to_string(),
        "--result-file".to_string(),
        result_file.to_string_lossy().to_string(),
    ];
    tmux(&command)?;

    let result_text = fs::read_to_string(&result_file)
        .map_err(|error| anyhow!("tmux-fuzzy-motion: popup did not produce result: {error}"))?;
    let result: InputResult = serde_json::from_str(&result_text)?;
    if let InputResult::Selected { target } = result {
        let target_pane = target
            .candidate
            .pane_id
            .clone()
            .unwrap_or_else(|| pane_id.clone());
        tmux_quiet(&s(&["select-pane", "-t", &target_pane]));
        if popup_state.state.scope == Scope::All
            && !popup_state
                .state
                .panes
                .iter()
                .any(|pane| pane.pane_id == target_pane && pane.in_copy_mode)
        {
            enter_copy_mode(&target_pane)?;
        }
        move_copy_cursor(&target_pane, &target)?;
    }
    command.clear();
    Ok(())
}

fn build_current_state(pane_id: &str, client_tty: &str) -> Result<PopupState> {
    let pane = get_pane_start_context(pane_id)?;
    if !pane.in_copy_mode {
        enter_copy_mode(pane_id)?;
    }
    let capture = fit_capture_to_height(capture_pane(pane_id)?, pane.height);
    Ok(PopupState {
        current_path: pane.current_path,
        x: None,
        y: None,
        state: InputState {
            scope: Scope::Current,
            pane_id: pane.pane_id,
            client_tty: client_tty.to_string(),
            display_lines: capture.display_lines,
            plain_lines: capture.lines,
            width: pane.width,
            height: pane.height,
            panes: Vec::new(),
        },
    })
}

fn build_all_pane_state(pane_id: &str, client_tty: &str) -> Result<PopupState> {
    let panes = list_window_panes(pane_id)?;
    let target = panes
        .iter()
        .find(|pane| pane.pane_id == pane_id)
        .or_else(|| panes.iter().find(|pane| pane.active))
        .or_else(|| panes.first())
        .ok_or_else(|| anyhow!("tmux-fuzzy-motion: pane not found"))?;
    let left = panes.iter().map(|pane| pane.left).min().unwrap_or(0);
    let top = panes.iter().map(|pane| pane.top).min().unwrap_or(0);
    let right = panes
        .iter()
        .map(|pane| pane.left + pane.width)
        .max()
        .unwrap_or(0);
    let bottom = panes
        .iter()
        .map(|pane| pane.top + pane.height)
        .max()
        .unwrap_or(0);
    let snapshots: Vec<PaneSnapshot> = thread::scope(|scope| {
        let handles = panes
            .iter()
            .map(|pane| {
                scope.spawn(move || {
                    let capture = fit_capture_to_height(capture_pane(&pane.pane_id)?, pane.height);
                    Ok(PaneSnapshot {
                        pane_id: pane.pane_id.clone(),
                        in_copy_mode: pane.in_copy_mode,
                        width: pane.width,
                        height: pane.height,
                        left: pane.left.saturating_sub(left),
                        top: pane.top.saturating_sub(top),
                        plain_lines: capture.lines,
                        display_lines: capture.display_lines,
                    })
                })
            })
            .collect::<Vec<_>>();
        handles
            .into_iter()
            .map(|handle| {
                handle
                    .join()
                    .map_err(|_| anyhow!("tmux-fuzzy-motion: failed to capture pane"))?
            })
            .collect::<Result<_>>()
    })?;
    let width = right.saturating_sub(left);
    let height = bottom.saturating_sub(top);
    let display_lines = compose_display_lines(&snapshots, width, height, &target.border_lines);
    Ok(PopupState {
        current_path: target.current_path.clone(),
        x: Some(build_popup_relative_position('x', left)),
        y: Some(build_popup_relative_position('y', top)),
        state: InputState {
            scope: Scope::All,
            pane_id: pane_id.to_string(),
            client_tty: client_tty.to_string(),
            display_lines,
            plain_lines: Vec::new(),
            width,
            height,
            panes: snapshots,
        },
    })
}

fn build_popup_relative_position(axis: char, target_origin: usize) -> String {
    if axis == 'x' {
        format!("#{{e|+|:#{{popup_pane_left}},#{{e|-|:{target_origin},#{{pane_left}}}}}}")
    } else {
        let top_offset =
            "#{?#{==:#{status-position},top},#{e|-|:#{client_height},#{window_height}},0}";
        let desired_top =
            format!("#{{e|+|:#{{e|-|:{target_origin},#{{window_offset_y}}}},{top_offset}}}");
        format!("#{{e|+|:#{{popup_height}},{desired_top}}}")
    }
}

fn compose_display_lines(
    panes: &[PaneSnapshot],
    width: usize,
    height: usize,
    border_lines: &str,
) -> Vec<String> {
    let mut rows = vec![vec![" ".to_string(); width]; height];
    let mut occupied = vec![vec![false; width]; height];
    for pane in panes {
        for row in pane.top..pane.top + pane.height {
            if let Some(occupied_row) = occupied.get_mut(row) {
                for col in pane.left..pane.left + pane.width {
                    if col < occupied_row.len() {
                        occupied_row[col] = true;
                    }
                }
            }
        }
    }
    for pane in panes {
        for (line_index, line) in pane.display_lines.iter().enumerate() {
            let Some(row) = rows.get_mut(pane.top + line_index) else {
                continue;
            };
            for (cell_index, cell) in compact_styled_display_cells(line).into_iter().enumerate() {
                let col = pane.left + cell_index;
                if col < row.len() {
                    row[col] = cell;
                }
            }
        }
    }
    draw_pane_borders(&mut rows, &occupied, border_lines);
    rows.into_iter()
        .map(|row| row.join("").trim_end().to_string())
        .collect()
}

fn border_set(border_lines: &str) -> (&'static str, &'static str, &'static str) {
    match border_lines {
        "single" => ("│", "─", "┼"),
        "double" => ("║", "═", "╬"),
        "heavy" => ("┃", "━", "╋"),
        "spaces" => (" ", " ", " "),
        _ => ("|", "-", "+"),
    }
}

fn draw_pane_borders(rows: &mut [Vec<String>], occupied: &[Vec<bool>], border_lines: &str) {
    let (vertical, horizontal, intersection) = border_set(border_lines);
    for row_index in 0..rows.len() {
        for column_index in 0..rows[row_index].len() {
            if occupied[row_index][column_index] {
                continue;
            }
            let left = column_index > 0 && occupied[row_index][column_index - 1];
            let right = occupied[row_index]
                .get(column_index + 1)
                .copied()
                .unwrap_or(false);
            let top = row_index > 0 && occupied[row_index - 1][column_index];
            let bottom = occupied
                .get(row_index + 1)
                .and_then(|row| row.get(column_index))
                .copied()
                .unwrap_or(false);
            let has_vertical = left || right;
            let has_horizontal = top || bottom;
            rows[row_index][column_index] = if has_vertical && has_horizontal {
                intersection.to_string()
            } else if has_vertical {
                vertical.to_string()
            } else if has_horizontal {
                horizontal.to_string()
            } else {
                rows[row_index][column_index].clone()
            };
        }
    }
}
